apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java

```diff
@@ -18,69 +18,118 @@
-                    spanTags.put(expression.getKey(), CustomizeExpression.parseExpression(expression.getValue(), context));
-                }
-            }
-            if (logs != null && !logs.isEmpty()) {
-                for (Map.Entry<String, String> entries : logs.entrySet()) {
-                    String expression = logs.get(entries.getKey());
-                    spanLogs.put(entries.getKey(), CustomizeExpression.parseExpression(expression, context));
+                    operationNameSuffix.append(CustomizeExpression.parseExpression(expression, evalContext));
                 }
             }
+            evalAndPopulate(evalContext, tags, spanTags);
+            evalAndPopulate(evalContext, logs, spanLogs);
+
             operationName = operationNameSuffix.insert(0, operationName).toString();
+            AbstractSpan localSpan = ContextManager.createLocalSpan(operationName);
 
-            AbstractSpan span = ContextManager.createLocalSpan(operationName);
-            if (!spanTags.isEmpty()) {
-                for (Map.Entry<String, String> tag : spanTags.entrySet()) {
-                    span.tag(Tags.ofKey(tag.getKey()), tag.getValue());
-                }
-            }
-            if (!spanLogs.isEmpty()) {
-                span.log(System.currentTimeMillis(), spanLogs);
-            }
+            tagSpanTags(localSpan, spanTags);
+            tagSpanLogs(localSpan, spanLogs);
+
+            spanTags.clear();
+            spanLogs.clear();
+
+            SpanDataHolder spanDataHolder = new SpanDataHolder(
+                localSpan, tags, logs, spanTags, spanLogs
+            );
+            miContext.setContext(spanDataHolder);
         }
     }
 
-    void afterMethod(Method method) {
-        ContextManager.stopSpan();
+    void afterMethod(Method method, Object ret, MethodInvocationContext miContext) {
+        if (!ContextManager.isActive()) {
+            return;
+        }
+        SpanDataHolder spanDataHolder = (SpanDataHolder) miContext.getContext();
+        if (spanDataHolder == null || spanDataHolder.localSpan == null) {
+            return;
+        }
+        AbstractSpan localSpan = spanDataHolder.localSpan;
+        if (ret == null) {
+            ContextManager.stopSpan(localSpan);
+            return;
+        }
+        Map<String, String> tags = spanDataHolder.tags;
+        Map<String, String> logs = spanDataHolder.logs;
+        Map<String, String> spanTags = spanDataHolder.spanTags;
+        Map<String, String> spanLogs = spanDataHolder.spanLogs;
+
+        Map<String, Object> evalContext = CustomizeExpression.evaluationReturnContext(ret);
+
+        evalReturnAndPopulate(evalContext, tags, spanTags);
+        evalReturnAndPopulate(evalContext, logs, spanLogs);
+
+        tagSpanTags(localSpan, spanTags);
+        tagSpanLogs(localSpan, spanLogs);
+
+        spanTags.clear();
+        spanLogs.clear();
+
+        ContextManager.stopSpan(localSpan);
```

Comment:
Please put `stopSpan` in the `try/final` block.

Reply:
Then in the `handleMethodException `, `if (ContextManager.isActive()) {` is not required anymore. And the span/context would not be leaked.

Reply:
changed